tools/kliopt/org/jetbrains/kliopt/KliOpt.kt

```diff
@@ -17,18 +17,23 @@
 
 package org.jetbrains.kliopt
 
-enum class DeprecatedState {
-    ALL {
-        override fun helpMessage(descriptor: Descriptor): String = "Warning: Option/argument -${descriptor.longName} is deprecated!"
-    },
-    ONLY_SHORT {
-        override fun helpMessage(descriptor: Descriptor): String =
-                "Warning: Short form -${(descriptor as? OptionDescriptor)?.shortName} of option -${descriptor.longName} is deprecated!"
-    },
-    ONLY_FULL {
-        override fun helpMessage(descriptor: Descriptor): String =
-                "Warning: Full form of option -${descriptor.longName} is deprecated!"
-    };
+sealed class DeprecatedState(val newName: String? = null) {
+    class AllForms(newName: String? = null) : DeprecatedState(newName) {
+        override fun helpMessage(descriptor: Descriptor) = "Warning: Option/argument -${descriptor.longName}" +
+                "${(descriptor as? OptionDescriptor)?.let { it.shortName?.let{ "-$it" } ?: ""} ?: ""} is deprecated!" +
+                "${newName?.let {" Use -$it instead."} ?: ""}"
+    }
+    class OnlyShort(newName: String? = null) : DeprecatedState(newName) {
+        override fun helpMessage(descriptor: Descriptor) =
+                "Warning: Short form -${(descriptor as? OptionDescriptor)?.shortName} of option " +
```

Comment:
Does user need to know that this is *short form* of some option? I don't think so. The only piece of information useful here is what option deprecated and how should it be replaced.

Reply:
It can be important, because if I read that option is deprecated I'd think that option is deprecated, not only it's short form.

Reply:
Indeed looks a bit too complicated for the problem solved. Maybe just introduce two separate options, one being deprecated, another is not, and only one is deprecated, and consumer takes concatenation of the two.

Reply:
I didn't understand approach you suggested.
Also I don't understand what is complicated here? Getting optional value in constructor?